### ISSUE TITLE
allow authors to "claim" datasets that they are named on as authors

### ISSUE THREAD

**Issue by only1chunts** [gigascience/gigadb-website#57]
in user profile page, allow authors to link their datasets to their user profile, this should also be linked to their Author ID in the author table so that those results are displayed in the search for that author ID. IDEALLY this would be done with ORCID, but that will cost us.

**Comment by rija** [comment]
Hi @only1chunts,


 For this task, I think we want to know and validate which user claim what dataset  using which author

Therefore the workflow should allow claims on a dataset author by users as well as  notification, validation/rejection of those claims. 

Do users who have submitted  a dataset need to be involved in some ways in the process if another user want to link himself/herself to an author of that dataset?

In other words, I’m wondering which of these workflows is preferred: 

* (A) ( ``claimant`` claims to be an author of a dataset ) —> ( ``submitter`` confirms authenticity ) -> ( ``admin``  validates claim by linking user and author )

* (B) ( ``claimant`` claims to be an author of a dataset ) —> ( ``admin``  validates claims by linking user and author )

* (C) ( ``claimant`` claims to be an author of a dataset ) —>  ( ``submitter``  validates claims by linking user and author )

* (D)  ( ``claimant`` claims to be an author of a dataset ) —> ( ``claimant``  link their user to the author )


I think :

(C) requires significantly more work to redesign the access control mechanism as submitter don’t have the right role to do that linkage.
(A) feels the most “correct” to me but it might be overkill from your perspective? It’s a more complex workflow than (B) too.
(D) feels dangerous and has same problem as (C) as in they would require admin role to perform the linkage.

**Comment by only1chunts** [comment]
Hi @rija , I thought I had already replied to this, but maybe that was via email instead of on here?! Anyway B is the correct option please.

**Comment by rija** [comment]
Thanks @only1chunts for the clarification,
I couldn't find the email thread it was discussed.

**Comment by rija** [comment]
Hi @only1chunts,

below are the user stories and suggested acceptance criteria I've drafted for this functionality

**From user's perspective:**

https://github.com/rija/gigadb-website/blob/claiming-dataset-57/tests/features/user-claims-dataset-57.feature

**From admin's perspective:**

https://github.com/rija/gigadb-website/blob/claiming-dataset-57/tests/features/admin-validates-dataset-claim-57.feature


Please, let me know whether they illustrate the requirements accurately or not.


As this functionality is asynchronous (user make a claim request at one point in time, admin validates it sometimes in the future),
I need to introduce a way to store the user request so that the user can create/reference it and the admins can get notification/take actions on it.

There are good architectural ways of solving this (e.g: message queues) but the easiest way is to implement it as a database table. Having not found an existing table that could fulfill that need, below is a suggested addition to the database schema to handle a commands that can be emitted by a user and actioned by another user (admin)

```
CREATE TABLE user_command (
    id SERIAL NOT NULL PRIMARY KEY,
    action_label VARCHAR(15) NOT NULL,
    requester_id INTEGER NOT NULL,
    actioner_id INTEGER,
    actionable_id INTEGER NOT NULL,
    request_date DATE,
    action_date DATE,
    status VARCHAR(15) NOT NULL,
  );
```

Typical records, using the same test data as for the acceptance criteria's scenario linked above,  would look like:

**Pending:**

| id 	| action_label 	| requester_id 	| actioner_id 	| actionable_id 	| request_date 			| action_date 			| status 	|
| ----- | ------------- | ------------- | ------------- | ----------------- | --------------------- | --------------------- | --------- |
| 100 	| claim_author	| 346 			| 				| 3791				| 2018-03-01 14:00T08:00| 						| pending	|


**Confirmed:**

| id 	| action_label 	| requester_id 	| actioner_id 	| actionable_id 	| request_date 			| action_date 			| status 	|
| ----- | ------------- | ------------- | ------------- | ----------------- | --------------------- | --------------------- | --------- |
| 100 	| claim_author	| 346 			| 		344		| 3791				| 2018-03-01 14:00T08:00| 2018-03-02 10:00T08:00| done		|

**Invalidated:**

| id 	| action_label 	| requester_id 	| actioner_id 	| actionable_id 	| request_date 			| action_date 			| status 	|
| ----- | ------------- | ------------- | ------------- | ----------------- | --------------------- | --------------------- | --------- |
| 100 	| claim_author	| 346 			| 		344		| 3791				| 2018-03-01 14:00T08:00| 2018-03-02 10:00T08:00| rejected	|

**Comment by only1chunts** [comment]
On "From user's perspective:"
https://github.com/rija/gigadb-website/blob/claiming-dataset-57/tests/features/user-claims-dataset-57.feature
i dont understand the last part?:
Scenario: a user already associated to an author cannot claim another author
	Given I sign in as a user
	And author "3794" is associated with user "345"
	When I am on "/dataset/100002"
	Then I should not see "Are you an author of this dataset? claim your dataset"

other than that they look fine.
As for the addition of a database table for tracking it. Can we use the same format as for all other table relationships? e.g. USER_AUTHOR contains the columns : 
ID
USER_ID
AUTHOR_ID
STATUS
it can include date fields if required (I cant think why they are needed?) not sure what the action_label column you suggested is for? or why you suggested 4 ID columns instead of 3?

**Comment by rija** [comment]
the **Scenario: a user already associated to an author cannot claim another author**
scenario is similar to 
**Scenario: Non logged-in visitors should not see the button**

If the user is already associated to an author, the button to claim a dataset authorship won't appear. 

It may be better to express it as follow:

```
Scenario: a user already associated to an author cannot claim another author
	Given author "3794" is associated with default user
	When I sign in as default user
        And I go to "/dataset/100002"
	Then I should not see "Are you an author of this dataset? claim your dataset"
```


Regarding the new table,
 
I was thinking more generically of a structure that can be used whenever the need for "user X request an action A to be performed on object Y at time T1, then action A is performed by admin Z at time T2" 

This table is not meant to hold persistent domain data, its a job/task queue implemented as a database table. It doesn't have to be specific to a domain entity like user/author, otherwise they will proliferate whenever a new asynchronous feature is needed.

**Comment by only1chunts** [comment]
@rija, I've done some testing of #56,#57 and #60 . Comments for all are listed here:
1 - Button on dataset page needs to be much smaller "Your dataset?" with hover-over for more description.
2 - button in admin pages to validate, reject and delete are invisible! only know where they are by the hover-over
3 - once validated the button to validate seems to remain in place?!
4 - deleting the "claim" from the admin pages makes no difference in the users page, i.e. it remains in their "authored datasets" list?
5 - in the users profile page the authored datasets table has an extra column on left called "Appears as" we dont need this column.
6 - we dont really need the column for operation in this table either since there is nothing the user can do at this point.
and I dont even remember what the "Subject" column was intended to display! so that can be removed from the authored datasets table (and probably from the submitted datasets table too)
7 - when selecting which author you are claiming to be, we should present the full author names and ORCIDs (if available), not just the display names.
8 - I made a claim, it was rejected. I tried to make a new claim and it tells me I cant because I have a pending claim already in progress?
9 - when updating the user info (/user/update/id/nnn) there is no mention of linked author, and no way to make the link. oddly in the view user info there is a link to author button?!
using the button in "view user" only1chunts@gmail.com I managed to add a random author as linked to the user, despite having ALREADY linked a different author to that user via the claim button? that claimed (and validated) author was not showing in the user view.

**Comment by rija** [comment]
@only1chunts,

Thank you for the feedback. I’ve reorganised my reply to your points in 3 categories with my comments inline in italic for ease of communication: 
a) Defects that need fixing (currently going through them)
b) Points that maybe defects, but I need more info from you
c) Points that may work, but I need to give you more details



## a) Defects that need fixing (in progress)

> 1 - Button on dataset page needs to be much smaller "Your dataset?" with hover-over for more description.

 _Done and committed to the branch._

> 5 - in the users profile page the authored datasets table has an extra column on left called "Appears as" we dont need this column.
> 6 - we dont really need the column for operation in this table either since there is nothing the user can do at this point.
> and I dont even remember what the "Subject" column was intended to display! so that can be removed from the authored datasets table (and probably from the submitted datasets table too)
> 7 - when selecting which author you are claiming to be, we should present the full author names and ORCIDs (if available), not just the display names.
> 8 - I made a claim, it was rejected. I tried to make a new claim and it tells me I cant because I have a pending claim already in progress?

_I'm working through them now_

> 9 - when updating the user info (/user/update/id/nnn) there is no mention of linked author, and no way to make the link. oddly in the view user info there is a link to author button?! 

_Ok, I’ll move the button from the view to the update form. See below in 10. my comment regarding the button behaviour_


## b) Points that maybe defects, but I need more info from you

> 2 - button in admin pages to validate, reject and delete are invisible! only know where they are by the hover-over

_I have tested in three web browsers (Chrome, Firefox, and Safari) on my Mac  and the validate, reject and delete icon-buttons appears correctly without having to hover over them. What environment are you testing on?_

_Also you need to make sure to recompile the LESS stylesheet files as I had to modify them to enable the icon on the buttons._ 

> 10. using the button in "view user" only1chunts@gmail.com I managed to add a random author as linked to the user, despite having ALREADY linked a different author to that user via the claim button? that claimed (and validated) author was not showing in the user view.

_I’ve tried to replicate that without success._ 
_After linking author and user with the claim mechanism, If I a go as an admin to the user view, the button is not there, instead there is a message as expected: This user is linked to author: Zhang G (3791)_

_The following two automated test scenarios already capture that situation and should prevent that issue to happen:_ 

```
@ok @javascript @admin-link-author-from-user
Scenario: On user view, if user has pending claim, link to pending claims
 	Given default admin user exists
 	And default user exists
 	And a user has a pending claim for author "3791"
	And I sign in as an admin
	When I go to "/user/view/id/345"
	Then the response should not contain "Link this user to an author"
	And the response should contain "This user has a pending claim. Click for details"

@ok @admin-link-author-from-user
Scenario: On user view, if user is already attached to an author, show author name
	Given default admin user exists
 	And default user exists
 	Given author "3794" is associated with user "345"
 	And I sign in as an admin
	When I go to "/user/view/id/345"
	Then the response should not contain "Link this user to an author"
	And the response should not contain "This user has a pending claim. Click for details"
	And the response should contain "Pan S"
```


## c) Points that may work, but I need to give you more details

> 3 - once validated the button to validate seems to remain in place?!
> 4 - deleting the "claim" from the admin pages makes no difference in the users page, i.e. it remains in their "authored datasets" list?

_The user claim system is just a job queue, it is not a domain object. It is kept for audit purpose, so deleting the claim record should not affect the author linking. The record should probably not actually be deleted unless a user cancel the claim (see 8. below)._

_Furthermore, the reason I kept the validate and reject button active is so that the linking and unlinking can be reverted as much as you want as long as the claim record is not deleted.  The button are idempotent, so they can be clicked again and again and will always attempt to do the same thing with no side effect._
_What do change though is whenever you click validate claim or reject a claim, the action date (which appears as a hover-over on the status) and the name of the admin will be updated (and the status will change too based on which button has been clicked)_ 

_So you if you want to remove an author linking, you don’t delete the record, but you click the reject button (and that action is reversible). The dataset is also then removed from the user profile._ 


Kind regards
Rija

**Comment by only1chunts** [comment]
Hi @rija 
**section A**
Thanks, I'll await your confirmation that those changes are ready to be looked at.

**section B:**
2 - @jessesiu installed it for me, so I'll get him to check the set up and that its not an artifact of the remote desktop application I used to access his machine.
10- I'll try again and see if I can reproduce it.

**section C:**
this is completely different behavior to everything else in GigaDB so is always going to be confusing. I think we should remove this table (either completely or just from from view if its still required behind the scenes). Instead, inline with all other GigaDB behavior, the claim should generate an email to database@giga with links to the relevant user admin page (/user/update/id/nnn) where the validate or reject buttons should be (still able to click multiple times to undo/redo)

**Comment by rija** [comment]
Hi @only1chunts,
I'll let you know when Section A is ready. 
Well noted for Section C, I'll redesign the workflow accordingly.

**Comment by rija** [comment]
Hi @only1chunts,

Section A is ready for re-testing.
Thank you.

**Comment by rija** [comment]
Hi @only1chunts,

I've now redone Section C:
Once a user successfully place a claim on a dataset, an email notification is sent to the configured admin_email, that contains the url to the user admin page where the admin can see a validate or reject button.  Once the claim is validated, the linking wil be shown on the user admin page, alongside an "unlink" button to undo the linking.

**Comment by only1chunts** [comment]
Hi @rija,
Sorry for the delay I tested the things Wednesday then forgot to actually write anything about them!
It all worked fine, thank you.
As a user I can make a claim on any dataset and it asks me which author I am. The system then emailed database@ to let admin know a user has made a claim. The link in that email took me to the relevant page to enable me to check the claim and validate or reject (I forgot to actually check what happens when I reject it). Validating the claim links the correct user to the requested author and blocks any further claims.
As admin I am able to link any author to any user (as long as that user doesn't already have a link author) via the admin pages.

I found no issues with any of the work, so as long as @pli888 and @jessesiu are happy with how the new code/database schema etc play with the current stuff I think this bit is done.

Thanks
Chris


----- Original Message -----
